src/hooks/useOnScrollReachedEndDetector/index.ts

```diff
@@ -0,0 +1,42 @@
+import React, { useCallback } from 'react';
+import { SCROLL_BUFFER } from '../../utils/consts';
+import { isAboutSame } from '../../modules/Channel/context/utils';
+import { useDebounce } from '../useDebounce';
+
+const BUFFER_DELAY = 500;
+
+export interface UseOnScrollReachedEndDetectorProps {
+  scrollRef: React.RefObject<HTMLDivElement>;
+  onReachedTop?: () => void;
+  onReachedBottom?: () => void;
+  onInBetween?: () => void;
+}
+
+export function useOnScrollPositionChangeDetector(props: UseOnScrollReachedEndDetectorProps): () => void {
+  const {
+    scrollRef,
+    onReachedTop,
+    onReachedBottom,
+    onInBetween,
+  } = props;
+
+  const cb = useCallback(() => {
+    const current = scrollRef?.current;
+    if (current) {
+      const {
+        scrollTop,
+        scrollHeight,
+        clientHeight,
+      } = current;
+      if (isAboutSame(scrollTop, 0, SCROLL_BUFFER)) {
+        onReachedTop();
+      } else if (isAboutSame(scrollHeight, clientHeight + scrollTop, SCROLL_BUFFER)) {
+        onReachedBottom();
+      } else {
+        onInBetween();
+      }
+    }
+  }, [scrollRef, onReachedTop, onReachedBottom, onInBetween]);
```

Comment:
```suggestion
import { usePreservedCallback } from "@sendbird/uikit-tools";


  const cb = usePreservedCallback(() => {
    const current = scrollRef?.current;
    if (current) {
      const {
        scrollTop,
        scrollHeight,
        clientHeight,
      } = current;
      if (isAboutSame(scrollTop, 0, SCROLL_BUFFER)) {
        onReachedTop();
      } else if (isAboutSame(scrollHeight, clientHeight + scrollTop, SCROLL_BUFFER)) {
        onReachedBottom();
      } else {
        onInBetween();
      }
    }
  });
```

Reply:
what is usePreservedCallback?

Reply:
Functions included in the dependency array must maintain their references, so they need to be wrapped with useCallback.
```tsx
const useMyHook = (callbackA, callbackB) => {
  return useCallback(() => {
    callbackA();
    callbackB();
  }, [callbackA, callbackB]);
}
```

When used like this, the use of useCallback is meaningless:
```tsx
const callback = useMyHook(() => { ... }, () => { ... });
```

Instead, you should use it like this, even though it makes the code complex and messy(useCallback hell):
```tsx
const callbackA = useCallback(() => { ... }, []);
const callbackB = useCallback(() => { ... }, []);
const callback = useMyHook(callbackA, callbackB);
```

By using usePreservedCallback, you can achieve the same result without worrying about these dependencies. The function references are always maintained consistently, and a new function is executed each time:
```tsx
const useMyHook = (callbackA, callbackB) => {
  return usePreservedCallback(() => {
    callbackA();
    callbackB();
  });
}

const callback = useMyHook(() => { ... }, () => { ... });
```

You can check the implementation here: https://github.com/sendbird/sendbird-uikit-core-ts/blob/main/packages/uikit-tools/src/hooks/usePreservedCallback.ts

Reply:
Thats pretty nice, thanks for this information. Ill apply this